### DISCUSSION TITLE
How to replace setExtras to provide metadata at message-level?

### DISCUSSION THREAD

**Discussion by dirtyhenry** [getsentry/sentry-cocoa#3285] [General]
I just realised today that `setExtra` had been deprecated in version 8.x.

I was using it extensively to attach metadata linked to events and messages, as opposed to the global state of the application, like so:

```swift
    func remoteLog(
        level: Logger.Level,
        message: Logger.Message,
        metadata: Logger.Metadata?
    ) {
        let prettyMetadata = ...transform metadata into a human-readable string...
        let userInfo = ["label": label, "metadata": prettyMetadata ?? "empty"]
        // Send to Sentry
        SentrySDK.capture(message: "\(message)") { scope in
            scope.setExtras(userInfo)
            scope.setLevel(level.sentryLevel)
        }
    }
```

I now read in [the docs](https://docs.sentry.io/platforms/apple/enriching-events/context/):

> **Additional Data is deprecated** in favor of structured contexts.

But my understanding of structured context is that it is relevant at a larger scale ie the application level, as opposed to just the message being captured.

I felt using extras had many benefits:

* Messages could be just messages, without any variable parts such as identifiers, etc., automatically providing issue grouping
* IDs were still very easily accessible for debugging and providing customer support

Am I missing something here? What should I do to replace my deprecated usage of `setExtras`? 🙏

**Comment by brustolin** [comment]
Hello @dirtyhenry 

You just need to use `setContext(value:key:)` instead

```swift
    func remoteLog(
        level: Logger.Level,
        message: Logger.Message,
        metadata: Logger.Metadata?
    ) {
        let prettyMetadata = ...transform metadata into a human-readable string...
        let userInfo = ["label": label, "metadata": prettyMetadata ?? "empty"]
        // Send to Sentry
        SentrySDK.capture(message: "\(message)") { scope in
            scope.setContext(value: userInfo, key: "userInfo")
            scope.setLevel(level.sentryLevel)
        }
    }
 ```
